### ISSUE TITLE
Error parsing trace db in carson/debugging 

### ISSUE THREAD

**Issue by stevenagy** [trailofbits/polytracker#6352]
Hey!

As requested, here's an issue I'm currently having when replicating the trace parsing script here (https://github.com/trailofbits/polytracker/tree/carson/debugging#running-and-analyzing-an-instrumented-program):
```
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(1, '__dummy__')
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.3.24-py3.8-linux-x86_64.egg/sqlalchemy/orm/query.py", line 3490, in one
    ret = self.one_or_none()
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.3.24-py3.8-linux-x86_64.egg/sqlalchemy/orm/query.py", line 3467, in one_or_none
    raise orm_exc.MultipleResultsFound(
sqlalchemy.orm.exc.MultipleResultsFound: Multiple rows were found for one_or_none()
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "parse.py", line 8, in <module>
    print(event)
  File "/mnt/bigdata/polytracker/polytracker/tracing.py", line 499, in __str__
    return f"{self.basic_block!s}#{self.entry_count()}"
  File "/mnt/bigdata/polytracker/polytracker/tracing.py", line 454, in entry_count
    event = self.previous_control_flow_event
  File "/mnt/bigdata/polytracker/polytracker/tracing.py", line 374, in previous_control_flow_event
    previous_event = self.previous_event
  File "/mnt/bigdata/polytracker/polytracker/database.py", line 323, in previous_event
    session.query(DBTraceEvent)
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.3.24-py3.8-linux-x86_64.egg/sqlalchemy/orm/query.py", line 3492, in one
    util.raise_(
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.3.24-py3.8-linux-x86_64.egg/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
sqlalchemy.orm.exc.MultipleResultsFound: Multiple rows were found for one()
```

The only PolyTracker-related change I've made is to add `fun:fgetc=custom` to `polytracker/build/share/polytracker/abi_lists/polytracker_abilist.txt`.
 
 Test program:
 ```
#include <stdio.h>
#include <stdlib.h> 

__attribute__((noinline))
void __dummy__(char x){
  if (x == NULL) return;
  return;
}

int main(int argc, char ** argv)
{
    FILE *fptr = fopen(argv[1], "r");
    if (fptr == NULL)
    {
        printf("Cannot open file \n");
        exit(0);
    }

    // Read contents from file
    char c = fgetc(fptr);
    while (c != EOF)
    {
        __dummy__(c);
        printf ("%c", c);
        c = fgetc(fptr);
    }

    fclose(fptr);
    return 0;
}
 ```

Example seed input: `ABCDEFGH` 

 Program args: `POLYFUNC=1 POLYTRACE=1 POLYPATH=seed ./test seed`

**Comment by ESultanik** [comment]
What did you do on the Python side to trigger that error? Just load the DB?

**Comment by stevenagy** [comment]
Loading it works fine, it's just iterating through the events where it crashes:
```
from polytracker import PolyTrackerTrace

trace = PolyTrackerTrace.load("polytracker.db")

for event in trace:
    print(event)
```

**Comment by ESultanik** [comment]
I was not able to reproduce this on commit 9fd55717f3b8ea6f590c6829c081a192ff1d025e running on Python 3.9.2 and sqlalchemy 1.3.23. I will try with sqlalchemy 1.3.24 next, which is what you appear to be using.

I added your example as a unit test, which you can run with:
```


Here is the output of iterating over the trace events:
```
DBFunctionEntry(0, 'main')
main@16#0
main@3#0
main@5#0
main@2#0
main@4#0
main@11#0
main@6#0
main@8#0
main@13#0
DBTaintAccess(10)
main@14#0
DBFunctionEntry(12, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(15)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(18)
main@12#0
main@15#0
main@17#0
main@13#1
DBTaintAccess(23)
main@14#1
DBFunctionEntry(25, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(28)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(31)
main@12#1
main@15#1
main@17#1
main@13#2
DBTaintAccess(36)
main@14#2
DBFunctionEntry(38, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(41)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(44)
main@12#2
main@15#2
main@17#2
main@13#3
DBTaintAccess(49)
main@14#3
DBFunctionEntry(51, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(54)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(57)
main@12#3
main@15#3
main@17#3
main@13#4
DBTaintAccess(62)
main@14#4
DBFunctionEntry(64, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(67)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(70)
main@12#4
main@15#4
main@17#4
main@13#5
DBTaintAccess(75)
main@14#5
DBFunctionEntry(77, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(80)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(83)
main@12#5
main@15#5
main@17#5
main@13#6
DBTaintAccess(88)
main@14#6
DBFunctionEntry(90, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(93)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(96)
main@12#6
main@15#6
main@17#6
main@13#7
DBTaintAccess(101)
main@14#7
DBFunctionEntry(103, '__dummy__')
__dummy__@2#0
__dummy__@4#0
DBTaintAccess(106)
__dummy__@0#0
__dummy__@3#0
DBFunctionReturn(109)
main@12#7
main@15#7
main@17#7
main@13#8
main@1#0
main@0#0
```

**Comment by ESultanik** [comment]
Sorry, I closed this by accident.

**Comment by stevenagy** [comment]
No worries! I'll pull and try on the latest version. 

Fwiw I'm using Python 3.8 and SQLAlchemy 1.3.23.dev0 (I couldn't find plain 1.3.23 so I modified `setup.py` to accept the dev0 version).

**Comment by ESultanik** [comment]
FWIW, we test against Python 3.8 in CI and it passed with the new `fgetc` unit test. I also just tried SQLAlchemy 1.3.24 and it seemed to work fine. So hopefully this is related to some other bug that we already fixed.

**Comment by stevenagy** [comment]
Interesting. Just installed SQLAlchemy 1.3.23 and I'm getting the same error:
```
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(0, 'main')
DBFunctionEntry(1, '__dummy__')
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3490, in one
    ret = self.one_or_none()
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3467, in one_or_none
    raise orm_exc.MultipleResultsFound(
sqlalchemy.orm.exc.MultipleResultsFound: Multiple rows were found for one_or_none()

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "parse.py", line 8, in <module>
    print(event)
  File "/mnt/bigdata/polytracker/polytracker/tracing.py", line 499, in __str__
    return f"{self.basic_block!s}#{self.entry_count()}"
  File "/mnt/bigdata/polytracker/polytracker/tracing.py", line 454, in entry_count
    event = self.previous_control_flow_event
  File "/mnt/bigdata/polytracker/polytracker/tracing.py", line 374, in previous_control_flow_event
    previous_event = self.previous_event
  File "/mnt/bigdata/polytracker/polytracker/database.py", line 323, in previous_event
    session.query(DBTraceEvent)
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3492, in one
    util.raise_(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
sqlalchemy.orm.exc.MultipleResultsFound: Multiple rows were found for one()
```

**Comment by ESultanik** [comment]
This will be resolved by #6353.
